### PR TITLE
feat: Default roles and RBAC permission checker.

### PR DESCRIPTION
**Pull request by ivarconr**
> **Note: This is a planned _Unleash Enterprise_ feature!**
> Parts of it will end up as part of the _Open-Source Distribution_. The more advanced control is only relevant for Enterprise. 

This PR Introduces first steps towards [RBAC according to our specifications](https://docs.google.com/document/d/1T2iuM9qgbgPfNMqXoQKEgdEG--a93lFzMsJHHuT0MHI). Rbac will assume users to exist in the Unleash user table with a unique id. This is required to make correct mappings between users and roles. 

Unleash will come with three "root" roles:

* **Admin** - Used to administer the Unleash instance. Is allowed to add/remove users, add them to roles and update role permissions. 
* **Regular** Represent users with typical read and write access to Unleash. They will typically be allowed create new projects, create feature toggles on the "default" project, etc. 
* **Read** Users only allowed to read data.

> We believe the *three root roles* will be part of Unleash Open-Source as well from v4.x release. At this point we can not promise exactly what will be Open-Source and Enterprise. 

In addition we plan to introduce two specific roles per project;
* **Admin** - Allowed to update the project. This includes adding and removing project members and their role. 
* **Regular** - Allowed to create and update feature toggles within the project. They can not update the project itself. 

In a later iteration we will introduce the concept of "custom roles". This will allow full customization to meet internal need of larger organisation. We need further investigation with customers before we land custom roles.

Related to #759

### CODE REVIEW

File: src/lib/permissions.js [hunk 1]
```diff
@@ -38,4 +40,6 @@ module.exports = {
     CREATE_ADDON,
     DELETE_ADDON,
     UPDATE_ADDON,
+    READ_ROLE,
```
Comment (ivarconr) [inline src/lib/permissions.js:43]: TODO: remove these for now.

File: src/lib/routes/admin-api/feature.js [hunk 2]
```diff
@@ -39,7 +39,7 @@ class FeatureController extends Controller {
         this.post('/:featureName/toggle/off', this.toggleOff, UPDATE_FEATURE);
         this.post('/:featureName/stale/on', this.staleOn, UPDATE_FEATURE);
         this.post('/:featureName/stale/off', this.staleOff, UPDATE_FEATURE);
-        this.get('/:featureName/tags', this.listTags, UPDATE_FEATURE);
+        this.get('/:featureName/tags', this.listTags);
```
Comment (ivarconr) [inline src/lib/routes/admin-api/feature.js:42]: @chriswk: It seemed like a bug that you must have the "UPDATE_FEATURE" rights in order to get tags for a feature toggle.

File: src/lib/routes/admin-api/rbac.js [hunk 3]
```diff
@@ -0,0 +1,98 @@
+'use strict';
+
+const Controller = require('../controller');
```
Comment (ivarconr) [inline src/lib/routes/admin-api/rbac.js:3]: TODO: remove this in the initial release (not needed before we address custom roles)

File: src/lib/services/access-service.ts [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+
+    /**
+     * Used to check if a user has access to the requested resource
+     * 
+     * @param user 
+     * @param permission 
+     * @param projectId 
+     */
+    async hasPermission(user: User, permission: string, projectId?: string): Promise<boolean> {
+        //TODO: require that project specific permissions specify projectId!
```
Comment (ivarconr) [inline src/lib/services/access-service.ts:93]: TODO: require that project specific permissions specify projectId!

File: src/lib/user.ts [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+            username,
+            imageUrl,
+            permissions,
+            seenAt,
+            loginAttempts,
+            createdAt,
+        }: UserData = { isAPI: false },
+    ) {
+        if (!username && !email) {
+            throw new TypeError('Username or Email us required');
```
Comment (ivarconr) [inline src/lib/user.ts:53]: fix typo

File: src/lib/db/access-store.ts [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+
+    async getRoleWithId(id: number): Promise<Role> {
+        return this.db
+            .select(['id', 'name', 'type', 'description'])
+            .where('id', id)
+            .first()
+            .from<Role>(T.ROLES);
+    }
+
+    async getRolesForProject(projectName: string): Promise<Role[]> {
```
Comment (ivarconr) [inline src/lib/db/access-store.ts:77]: change to `projectId`

File: src/lib/db/feature-toggle-store.js [hunk 7]
```diff
@@ -98,6 +98,14 @@ class FeatureToggleStore {
             .then(this.rowToFeature);
     }
 
+    async getProjectId(name) {
```
Comment (ivarconr) [inline src/lib/db/feature-toggle-store.js:101]: Verify that this does not throw if feature toggle is missing.-

File: src/lib/services/project-service.ts [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+    ): Promise<void> {
+        const roles = await this.accessService.getRolesForProject(projectId);
+        const role = roles.find(r => r.id === roleId);
+        if (!role) {
+            throw new NotFoundError(
+                `Couldn't find roleId=${roleId} on project=${projectId}`,
+            );
+        }
+
+        if (role.type === 'project-admin') {
```
Comment (ivarconr) [inline src/lib/services/project-service.ts:184]: use enum.

File: src/test/e2e/helpers/database-init.js [hunk 9]
```diff
@@ -28,6 +28,10 @@ async function resetDatabase(stores) {
         stores.db('tags').del(),
         stores.db('tag_types').del(),
         stores.db('addons').del(),
+        stores.db('users').del(),
+        // stores.db('roles').del(),
```
Comment (ivarconr) [inline src/test/e2e/helpers/database-init.js:32]: fix:

File: src/lib/routes/admin-api/feature.js (hunk 2 above)
Comment (chriswk) [reply src/lib/routes/admin-api/feature.js:42]: Yes. Absolutely. Good catch

File: src/migrations/20210217195834-rbac-tables.js [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+      (
+          role_id     integer not null references roles (id) ON DELETE CASCADE,
+          project     text,
+          permission  text not null,
+          created_at  TIMESTAMP WITH TIME ZONE DEFAULT now()
+      );
+    
+      WITH admin AS (
+        INSERT INTO roles(name, description, type)
+        VALUES ('Admin', 'Can read and write all resources in Unleash. This includes add and remove users and roles.', 'root')
```
Comment (ivarconr) [inline src/migrations/20210217195834-rbac-tables.js:29]: @FredrikOseberg I would love your feedback on these Role descriptions. They will typically be shown in the UI as helpers.

File: src/lib/user.ts [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+        }
+        Joi.assert(email, Joi.string().email(), 'Email');
+        Joi.assert(username, Joi.string(), 'Username');
+        Joi.assert(name, Joi.string(), 'Name');
+
+        this.id = id;
+        this.isAPI = isAPI;
```
Comment (ivarconr) [inline src/lib/user.ts:60]: @chriswk this feels a bit confusing. Suggestion on how to differentiate a "API user" from a "human user"?

File: src/lib/middleware/rbac-middleware.test.ts [hunk 12]
```diff
@@ -0,0 +1,257 @@
...
+    };
+
+    func(req, undefined, cb);
+
+    const hasAccess = await req.checkRbac(perms.ADMIN);
+
+    t.false(hasAccess);
+});
+
+test('should not return false for missing user', async t => {
```
Comment (chriswk) [inline src/lib/middleware/rbac-middleware.test.ts:124]: Should return false?

Comment (ivarconr) [reply src/lib/middleware/rbac-middleware.test.ts:124]: indeed!

File: src/lib/middleware/rbac-middleware.ts [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+            // Temporary workaround to figure our projectId for feature toggle updates.
+            if ([UPDATE_FEATURE, DELETE_FEATURE].includes(permission)) {
+                const { featureName } = params;
+                projectId = await featureToggleStore.getProjectId(featureName);
+            } else if (permission === CREATE_FEATURE) {
+                projectId = req.body.project;
+            }
+
+            return accessService.hasPermission(user, permission, projectId);
+        };
```
Comment (chriswk) [inline src/lib/middleware/rbac-middleware.ts:47]: I need a double-take when I read the `if (!user || !user.id)` check. Seeing as how you've already checked for user on L26, may I suggest something like this?
```suggestion
        req.checkRbac = async (permission: string) => {
            const { user, params } = req;

            // Support ADMIN API tokens for enterpriseAuthentication.
            if (user) {
                if(user.isAPI) {
                 return user.permissions.includes(ADMIN);
                }
                if (!user.id) {
                  logger.error('RBAC requires users to have a unique id.');
                }
               // For /api/admin/projects/:projectId we will find it as part of params
               let { projectId } = params;

               // Temporary workaround to figure our projectId for feature toggle updates.
               if ([UPDATE_FEATURE, DELETE_FEATURE].includes(permission)) {
                   const { featureName } = params;
                   projectId = await featureToggleStore.getProjectId(featureName);
               } else if (permission === CREATE_FEATURE) {
                  projectId = req.body.project;
               }
               return accessService.hasPermission(user, permission, projectId);    
            } 
            logger.error('RBAC requires a user');
            return false; 
        };
```

File: src/lib/user.ts (hunk 5 above)
Comment (ivarconr) [reply src/lib/user.ts:53]: fixed!

File: src/lib/user.ts (hunk 11 above)
Comment (chriswk) [reply src/lib/user.ts:60]: Having a boolean column in the user table for whether or not it's an API user doesn't seem that complicated. 
Or maybe have a separate class/type/table for ApiUsers, that complicates join operations though, but strictly speaking they are different concepts?

File: src/migrations/20210218090213-generate-server-identifier.js [hunk 14]
```diff
@@ -15,7 +15,7 @@ exports.up = function(db, cb) {
 exports.down = function(db, cb) {
     db.runSql(
         `
-        DROP FROM settings WHERE name = 'instanceInfo'
+        DELETE FROM settings WHERE name = 'instanceInfo'
```
Comment (chriswk) [inline src/migrations/20210218090213-generate-server-identifier.js:18]: This is fixed on master :)

File: src/test/e2e/helpers/database-init.js (hunk 9 above)
Comment (chriswk) [reply src/test/e2e/helpers/database-init.js:32]: Consider using db-migrator's down function here?

File: src/lib/user.test.js [hunk 15]
```diff
@@ -38,7 +38,7 @@ test('should require email or username', t => {
         { instanceOf: Error },
     );
 
-    t.is(error.message, 'Username or Email us reuqired');
+    t.is(error.message, 'Username or Email us required');
```
Comment (chriswk) [inline src/lib/user.test.js]:
```suggestion
    t.is(error.message, 'Username or Email is required');
```

File: src/migrations/20210218090213-generate-server-identifier.js (hunk 14 above)
Comment (ivarconr) [reply src/migrations/20210218090213-generate-server-identifier.js:18]: ok, i will rebase :D

File: src/lib/db/feature-toggle-store.js (hunk 7 above)
Comment (ivarconr) [reply src/lib/db/feature-toggle-store.js:101]: needed to add a small check.

File: src/lib/user.ts (hunk 11 above)
Comment (ivarconr) [reply src/lib/user.ts:60]: lets keep it until we have done the api tokens thing in #755

File: src/lib/middleware/rbac-middleware.ts (hunk 13 above)
Comment (ivarconr) [reply src/lib/middleware/rbac-middleware.ts]: I have decided a bit simpler approach. First check that the user exists. Then it is safe to assume it exist later on.

File: src/lib/middleware/rbac-middleware.ts [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+        return next();
+    };
```
Comment (chriswk) [inline src/lib/middleware/rbac-middleware.ts:54]: Even better. :)

**Review by chriswk — APPROVED** [review]

File: src/lib/db/access-store.ts [hunk 17]
```diff
@@ -0,0 +1,163 @@
...
+import metricsHelper from '../metrics-helper';
+import { DB_TIME } from '../events';
+
+const T = {
+    ROLE_USER: 'role_user',
+    ROLES: 'roles',
+    ROLE_PERMISSION: 'role_permission',
+};
+
+export interface Permission {
```
Comment (FredrikOseberg) [inline src/lib/db/access-store.ts:12]: We haven't discussed this, but what do you think about a convention for naming interfaces? One quite common convention is to prefix interfaces with I. In this instance the interface would be IPermission. The benefit is that it clearly communicates in the code what it represents. Most of the time you can deduce it from the use context, but there's still some cognitive load involved.

File: src/lib/db/project-store.js [hunk 18]
```diff
@@ -51,11 +51,10 @@ class ProjectStore {
     }
 
     async create(project) {
-        await this.db(TABLE)
+        const [id] = await this.db(TABLE)
```
Comment (FredrikOseberg) [inline src/lib/db/project-store.js:54]: Is catching no longer relevant here?

File: src/lib/middleware/rbac-middleware.ts (hunk 16 above)
Comment (FredrikOseberg) [reply src/lib/middleware/rbac-middleware.ts:54]: Elegant solve!

File: src/lib/services/access-service.ts [hunk 19]
```diff
@@ -0,0 +1,219 @@
...
+    imageUrl?: string;
+}
+
+interface RoleData {
+    role: Role;
+    users: User[];
+    permissions: Permission[];
+}
+
+interface IPermission {
```
Comment (FredrikOseberg) [inline src/lib/services/access-service.ts:41]: Ah, here we're using IPermission. Should we be consistent one way or the other?

File: src/lib/services/access-service.ts [hunk 20]
```diff
@@ -0,0 +1,219 @@
...
+        return this.permissions;
+    }
+
+    async addUserToRole(userId: number, roleId: number) {
+        return this.store.addUserToRole(userId, roleId);
+    }
+
+    async setUserRootRole(userId: number, roleName: RoleName ) {
+        const userRoles = await this.store.getRolesForUserId(userId);
+        const currentRootRoles = userRoles.filter(r => r.type === 'root');
```
Comment (FredrikOseberg) [inline src/lib/services/access-service.ts:110]: Should 'root' be a constant?

File: src/lib/services/access-service.ts [hunk 21]
```diff
@@ -0,0 +1,219 @@
...
+    async createDefaultProjectRoles(owner: User, projectId: string) {
+        if(!projectId) {
+            throw new Error("ProjectId cannot be empty");
+        }
+
+        const adminRole = await this.store.createRole(
+            RoleName.ADMIN,
+            RoleType.PROJECT,
+            projectId,
+            `Admin role for project "${projectId}"`,
```
Comment (FredrikOseberg) [inline src/lib/services/access-service.ts:188]: Do these override the default descriptions in the migration?

Comment (FredrikOseberg) [reply src/lib/services/access-service.ts:188]: Description suggestions: 

Project Admin
Users with the project admin role have full control over the project, and can add and manage other users within the project context, manage feature toggles within the project, and control advanced project features like archiving and deleting the project.

Project Regular
Users with the regular role within a project are allowed to view, create and update feature toggles, but have limited permissions in regards to managing the projects user access and can not archive or delete the project. 

Global Admin: 
Users with the global admin role have superuser access to Unleash and can perform any operation within the unleash platform.

Global Regular:
Users with the global regular role have access most features in Unleash, but can not manage users and roles in the global scope. If a user with a global regular role creates a project, they will become a project admin and receive superuser rights within the context of that project.

File: src/lib/services/project-service.ts [hunk 22]
```diff
@@ -0,0 +1,195 @@
...
+
+        await this.eventStore.store({
+            type: eventType.PROJECT_UPDATED,
+            createdBy: getCreatedBy(user),
+            data: project,
+        });
+    }
+
+    async deleteProject(id: string, user: User): Promise<void> {
+        if (id === 'default') {
```
Comment (FredrikOseberg) [inline src/lib/services/project-service.ts:87]: Should this be a constant?

**Review by FredrikOseberg — APPROVED** [review]
Overall I think this is a quite elegant solution. I especially like the middleware implementation, which makes for a really clean interface. I've added some questions throughout the code regarding minor things. Nice job!

File: src/lib/services/access-service.ts (hunk 20 above)
Comment (chriswk) [reply src/lib/services/access-service.ts:110]: Alternately part of an enum?

File: src/lib/db/project-store.js (hunk 18 above)
Comment (ivarconr) [reply src/lib/db/project-store.js:54]: Not really. This was actually a regression I haven't spotted before touching this code again. Swallowing these throws is kind of bad for the user.

File: src/lib/db/access-store.ts (hunk 17 above)
Comment (ivarconr) [reply src/lib/db/access-store.ts:12]: Let's try that. I will refactor all interfaces to "**I**MyInterface"

File: src/lib/services/access-service.ts (hunk 21 above)
Comment (ivarconr) [reply src/lib/services/access-service.ts:188]: The migration only takes care of the default "root roles". The project specific ones are created when a new project is created.

File: src/lib/services/access-service.ts (hunk 20 above)
Comment (ivarconr) [reply src/lib/services/access-service.ts:110]: there is already a enum for RoleType

**Pull request merged**